### PR TITLE
feat: alpine based docker image

### PR DESCRIPTION
**Pull request by tynes**
## Motivation

I have tasks that need to run in the cloud where bringing up a K8s pod
that I can attach to with `cast` would be nice. A docker image is required
for this and I'm sure other people would want a docker image so putting
the dockerfile here makes sense.

## Solution

This commit creates a docker image that includes
both `cast` and `forge`.

It builds off of https://github.com/gakonst/foundry/pull/914.

The image comes out to 26.4MB.

```bash
/ # du -h /usr/local/bin/forge
12.8M   /usr/local/bin/forge
/ # du -h /usr/local/bin/cast
7.1M    /usr/local/bin/cast
```

Example usage:

```bash
$ docker run --rm --entrypoint cast foundry:latest block --rpc-url https://mainnet.optimism.io latest
$ docker run --rm foundry:latest 'cast block --rpc-url https://mainnet.optimism.io latest'
```

Co-authored-by: Abdul Rabbani

---

One question is whether or not the container should include `bash` - some people might want to mount in bash scripts to run. I personally think yes but did not include `bash` in the image as to keep it minimal, will update if there is consensus around this.

### CODE REVIEW

**Review by onbjerg — COMMENTED** [review]
You can further shrink the image if you run `strip /usr/src/foundry/target/release/cast` and `strip /usr/src/foundry/target/release/forge`

Also if this builds off of #914 let's merge into that branch instead :)

**Comment by tynes** [comment]
> You can further shrink the image if you run `strip /usr/src/foundry/target/release/cast` and `strip /usr/src/foundry/target/release/forge`
> 
> Also if this builds off of #914 let's merge into that branch instead :)

Good looks on using `strip`, that significantly made the image smaller.

I opened a PR that doesn't merge into #914 because the author said they don't have time to contribute more, see https://github.com/gakonst/foundry/pull/914/files#r830395267. If you are strongly opinionated on me opening this PR against their repo, I can.

File: Dockerfile
Comment (dmfxyz) [inline Dockerfile:18]: Will targeting native cpu instruction set require a multi-architecture docker build when this images becomes part of a CI/CD pipeline?

**Comment by dmfxyz** [comment]
@tynes running into the classic alpine solc issues when trying to compile with forge in the image. Are you seeing this? Could just be an issue on my local.

```
$> docker build -t ercc721 .
Sending build context to Docker daemon  26.03MB
Step 1/3 : from foundry-alpine-tynes
 ---> 25502cb6ddcd
Step 2/3 : COPY . .
 ---> ec207d41b853
Step 3/3 : RUN forge test -vv
 ---> Running in 021765ffee03
Error:
   0: "/root/.svm/0.8.13/solc-0.8.13": No such file or directory (os error 2)

Location:
   cli/src/cmd/mod.rs:96

Backtrace omitted.
Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
The command '/bin/sh -c forge test -vv' returned a non-zero code: 1
$>
```

**Comment by tynes** [comment]
> @tynes running into the classic alpine solc issues when trying to compile with forge in the image. Are you seeing this? Could just be an issue on my local.
> 
> ```
> $> docker build -t ercc721 .
> Sending build context to Docker daemon  26.03MB
> Step 1/3 : from foundry-alpine-tynes
>  ---> 25502cb6ddcd
> Step 2/3 : COPY . .
>  ---> ec207d41b853
> Step 3/3 : RUN forge test -vv
>  ---> Running in 021765ffee03
> Error:
>    0: "/root/.svm/0.8.13/solc-0.8.13": No such file or directory (os error 2)
> 
> Location:
>    cli/src/cmd/mod.rs:96
> 
> Backtrace omitted.
> Run with RUST_BACKTRACE=1 environment variable to display it.
> Run with RUST_BACKTRACE=full to include source snippets.
> The command '/bin/sh -c forge test -vv' returned a non-zero code: 1
> $>
> ```

Thanks for pointing this out. It turns out that `/lib64/ld-linux-x86-64.so.2` didn't exist, which can be included with the package [libc6-compat](https://pkgs.alpinelinux.org/package/edge/main/x86_64/libc6-compat)

```
/project # file ~/.svm/0.8.13/solc-0.8.13
/root/.svm/0.8.13/solc-0.8.13: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 3.2.0, BuildID[sha1]=acd57a71be05049f965860e3311f82154d39c5ea, stripped
```

After installing that package, a new error happens

```
/project # ~/.svm/0.8.13/solc-0.8.13
Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /root/.svm/0.8.13/solc-0.8.13)
Error relocating /root/.svm/0.8.13/solc-0.8.13: __vsnprintf_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __memmove_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __memcpy_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __memset_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __snprintf_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: dlmopen: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __sprintf_chk: symbol not found
Error relocating /root/.svm/0.8.13/solc-0.8.13: __strftime_l: symbol not found
```

Perhaps we need a `solc` that is built against musl instead of glibc? Or we need to install the correct packages

`solc` can run on alpine fwiw
https://github.com/ethereum/solidity/blob/44fdcdcb12d72383c3dce2ff4e623616c07143f7/scripts/Dockerfile

**Comment by dmfxyz** [comment]
@tynes take a look at https://github.com/abdulrabbani00/foundry/pull/3/files. Been taking a look at this for a couple nights. I opened this PR in order to build off of #914. I think it gets the job done but def needs more eyes and testing.

**Comment by dmfxyz** [comment]
If we are migrating to this PR happy to open a PR against your fork as well :)

**Comment by tynes** [comment]
> @tynes take a look at https://github.com/abdulrabbani00/foundry/pull/3/files. Been taking a look at this for a couple nights. I opened this PR in order to build off of #914. I think it gets the job done but def needs more eyes and testing.

Awesome, thanks for looking into this problem! I'm gonna be mostly afk this weekend so feel free to coordinate with the other PR or open your own if you are around this weekend. Happy to help test/do whatever to get this over the line :)

**Review by gakonst — APPROVED** [review]
Supportive, defer to @mattsse @onbjerg on any extra changes

**Comment by dmfxyz** [comment]
> Awesome, thanks for looking into this problem! I'm gonna be mostly afk this weekend so feel free to coordinate with the other PR or open your own if you are around this weekend. Happy to help test/do whatever to get this over the line :)

@tynes opened a PR into your branch. https://github.com/tynes/dapptools-rs/pull/1. This dockerfile should resolve the issues we saw above with solc (tested on a number of platforms). If you rebase or merge that in than we can continue using this PR (or I can open a new one if you'd prefer that and mark you as co-author if that's less of a burden).

File: Dockerfile
Comment (huitseeker) [inline Dockerfile:21]: Rust has built-in stripping since [1.59](https://blog.rust-lang.org/2022/02/24/Rust-1.59.0.html#creating-stripped-binaries), could that be used instead?

File: Dockerfile
Comment (mattsse) [inline Dockerfile:18]: not sure about this one, this may be fine when built for your own PC but probably not when push to ghcr or dockerhub?

File: Dockerfile
Comment (mattsse) [reply Dockerfile:21]: TIL,
perhaps we add a `stripped` profile that `inherits` from `release` and sets `strip`

File: Dockerfile
Comment (gakonst) [reply Dockerfile:18]: Good Q. I am not sure. I think let's remove.

Comment (gakonst) [inline Dockerfile:18]:
```suggestion
```

File: Dockerfile
Comment (gakonst) [reply Dockerfile:21]: Defer to whatever @tynes wants here, I think they're equivalent?

**Review by gakonst — COMMENTED** [review]
LGTM so far. I'd love if we could also have a CI integration here, but acknowledge that it might be too much work. Up to you @tynes

File: Dockerfile
Comment (dmfxyz) [reply Dockerfile:18]: Yeah mentioned concerns with this above. I think it should only be used for local installs.

**Comment by dmfxyz** [comment]
> LGTM so far. I'd love if we could also have a CI integration here, but acknowledge that it might be too much work. Up to you @tynes @dmfxyz

I think what might be best is to merge a Dockerfile first (but don't merge this PR yet, as there are outstanding changes needed here https://github.com/tynes/dapptools-rs/pull/1), and then open a separate PR for CI integration. We have 2 PRs (three if I were to open one rather than PRing into @tynes branch) for Docker right now. I say we get them closed / merged and we can start from a clean slate for CI integration.

**Comment by onbjerg** [comment]
Let's get https://github.com/tynes/dapptools-rs/pull/1 merged into this PR and get this over the finish line. It would be nice if we could get a workflow of some sort (see #914 for inspiration) to build a Docker container nightly and push it to GHCR :smile:

**Comment by dmfxyz** [comment]
> Let's get [tynes#1](https://github.com/tynes/dapptools-rs/pull/1) merged into this PR and get this over the finish line. It would be nice if we could get a workflow of some sort (see #914 for inspiration) to build a Docker container nightly and push it to GHCR 😄

Yeah, I have an example workflow [here](https://github.com/dmfxyz/foundry/blob/dockerize/.github/workflows/docker-publish.yml), and a successful run and publish [here](: https://github.com/dmfxyz/foundry/runs/5615229053?check_suite_focus=true). But it needs some more refinement and won't be ready for a bit.

**Review by gakonst — APPROVED** [review]
lgtm! let's ship this and we can do the GHA release separately

**Pull request merged**
